compiler/rustc_target/src/spec/mod.rs

```diff
@@ -1014,6 +1014,8 @@ supported_targets! {
     ("armv7-unknown-linux-uclibceabihf", armv7_unknown_linux_uclibceabihf),
 
     ("x86_64-unknown-none", x86_64_unknown_none),
+
+    ("mips64-openwrt-linux-musl", mips64_openwrt_linux_musl),
```

Comment:
Shouldn't this target be `*-*-*-muslabi64`?

Reply:
From the internal `mips64-unknown-linux-musl` tuple:

```
        // LLVM doesn't recognize "muslabi64" yet.
        llvm_target: "mips64-unknown-linux-musl".to_string(),
```

OpenWrt reports as `mips64_openwrt_linux_musl` in this case.

Reply:
Tier 3 - mips64-openwrt-linux-musl

```
    A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)
```

I would be responsible for maintaining the `mips64-openwrt-linux-musl` tuple

```
    Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

    Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.
```

The `mips64-openwrt-linux-musl` conforms to existing tuple naming conventions.

    Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.
        The target must not introduce license incompatibilities.
        Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).
        The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.
        If the target supports building host tools (such as rustc or cargo), those host tools must not depend on proprietary (non-FOSS) libraries, other than ordinary runtime libraries supplied by the platform and commonly used by other binaries built for the target. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.
        Targets should not require proprietary (non-FOSS) components to link a functional binary or library.
        "onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.
```

OpenWrt is released under GPLv2 (https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=LICENSES/GPL-2.0)

```
    Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.
        This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.
```

This PR constitutes no binding agreement between myself and the `rust-lang` project.  I am not entitled to make any representation on behalf of, or in respect to, the OpenWrt project.

```
    Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.
```

The goal is to introduce mainline maintained tuples to the extent possible.

```
    The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running tests (even if they do not pass), the documentation must explain how to run tests for the target, using emulation if possible or dedicated hardware if necessary.
```

Compilation of the `mips64-openwrt-linux-musl` toolchain is done completely within the OpenWrt build environment.

```
    Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.

    Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.
        In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.
```

No burden on other toolchains is foreseen.

```
If a tier 3 target stops meeting these requirements, or the target maintainers no longer have interest or time, or the target shows no signs of activity and has not built for some time, or removing the target would improve the quality of the Rust codebase, we may post a PR to remove it; any such PR will be CCed to the target maintainers (and potentially other people who have previously worked on the target), to check potential interest in improving the situation.
```

Removal of the Tier 3 tuple, if abandoned, is preferred to non-functional.

Reply:
The relationship between Rust's target names and LLVM target names is not particularly correlated, in practice.

Given that this target both sets `cfg(abi = abi64)` and that the other Rust's MIPS64 targets are all `muslabi64`, I feel like there would need to be a pretty good reason to not follow the same convention for this target.

Reply:
I am using the internal `$(REAL_TARGET_GNU_NAME)` as defined by the OpenWrt Build system (https://github.com/openwrt/openwrt/blob/87d489f67a35968242cd4f192dc363dc87056844/rules.mk#L122)

I will ask if it SHOULD be reporting as `musl` or `muslabi64`.  MIPS64 under OpenWrt is -N64 and not -O32

Reply:
I'm still looking for an answer either way.  I'd rather do this correctly now than have to deal with later.

Reply:
What was the outcome here?

Reply:
Wonder why it didn't take my comment..

LLVM doesn't support `muslabi64`, only `musl`.  `rust-lang` redirects the target because muslabi64 doesn't actually exist for mips64 (check the other MUSL 64-bit targets, like aarch64 or ppc64).

I would prefer to keep it `musl`, as that is what OpenWrt uses for it's build system.  When LLVM updates the mips64 definitions to `muslabi64`, I'll update the OpenWrt tuples to match.

Reply:
> LLVM doesn't support muslabi64

Just reiterating – what triples LLVM does or does not support doesn't really matter much for the purposes of `rustc`'s target naming. The two triples aren't required to match and usually do not.

> I would prefer to keep it musl, as that is what OpenWrt uses for it's build system.

Cool, works for me.